workflow-ui/backstack-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateFrame.kt

```diff
@@ -1,19 +1,179 @@
+ *
+ * This class manages both the Android view state traditionally saved and restored through the
+ * [View.onSaveInstanceState]/[View.onRestoreInstanceState], as well as the modern
+ * [SavedStateRegistry] for the view exposed via [ViewTreeSavedStateRegistryOwner].
+ *
+ * A frame can either be [Detached] if it's simply holding the saved data for a hidden view, or
+ * [Attached] when it's managing a visible view.
  */
-internal data class ViewStateFrame(
+@OptIn(WorkflowUiExperimentalApi::class)
+internal class ViewStateFrame(
   val key: String,
-  val viewState: SparseArray<Parcelable>
-) : Parcelable {
+  @VisibleForTesting(otherwise = PRIVATE)
+  var viewState: SparseArray<Parcelable>?,
+  @VisibleForTesting(otherwise = PRIVATE)
+  var androidXBundle: Bundle?
+) : Parcelable, SavedStateRegistryOwner, WorkflowLifecycleOwner {
+  constructor(key: String) : this(key, null, null)
+
+  private sealed class AttachState {
+    /**
+     * The frame is hidden and not attached to any view.
+     * In this state the frame is simply a [Parcelable] holder for state data to be restored later.
+     */
+    object Detached : AttachState()
+
+    /**
+     * The frame represents a screen that is actively being shown.
+     * In this state the frame may be asked to restore and/or save the current view's state.
+     */
+    class Attached(
+      val view: View,
+      val lifecycleOwner: WorkflowLifecycleOwner,
+      val savedStateController: SavedStateRegistryController
+    ) : AttachState()
+  }
+
+  private var attachState: AttachState = Detached
+
+  override fun getLifecycle(): Lifecycle =
+    requireAttached().lifecycleOwner.lifecycle
+
+  override fun getSavedStateRegistry(): SavedStateRegistry =
+    requireAttached().savedStateController.savedStateRegistry
+
+  /**
+   * Sets up any tags on the [view] that need to point to state or lifecycle information managed
+   * by this frame.
+   *
+   * Note that this method requires [WorkflowLifecycleOwner.installOn] to already have been called
+   * for the view. It will read the [WorkflowLifecycleOwner] and then replace it with this
+   * [ViewStateFrame].
+   */
+  fun attach(view: View) {
+    require(attachState is Detached) {
+      "Expected attach to only be called once per view."
+    }
+    attachState = Attached(
+      view = view,
+      lifecycleOwner = requireNotNull(WorkflowLifecycleOwner.get(view)) {
+        "Expected BackStackContainer screen view to have a WorkflowLifecycleOwner tag set."
+      },
+      savedStateController = SavedStateRegistryController.create(this)
+    )
+
+    lifecycle.addObserver(object : LifecycleEventObserver {
+      override fun onStateChanged(
+        source: LifecycleOwner,
+        event: Event
+      ) {
+        if (event == ON_DESTROY) {
+          changeToDetachedState()
+        }
+      }
+    })
+
+    ViewTreeSavedStateRegistryOwner.set(view, this)
+    ViewTreeLifecycleOwner.set(view, this)
```

Comment:
Notice the parallels with [this](https://android-review.googlesource.com/c/platform/frameworks/support/+/1577946).